website/src/components/Stats/Stats.components.tsx

```diff
@@ -38,30 +39,208 @@ type MessageTreeStateStatsProps = {
   };
 };
 
+const findLangRow = (rows: string[][], lang: string): number => {
+  for (let i = 0; i < rows.length; i++) {
+    const row_lang = rows[i][0];
+    if (row_lang === lang) return i;
+  }
+
+  return -1;
+};
+
+function getTableData(
+  stat: StatType & {
+    stats: {
+      lang: string;
+      state: string;
+      count: number;
+    }[];
+  },
+  t
+) {
+  // The headers come from the message tree State.
+  // These two states are not considered because they make
+  // the table too wide and they are not so interesting:
+  // - "ready_for_scoring",
+  // - "scoring_failed"
+  let headers = [
+    "language", // This one doesn't come from the message tree state
+    "prompt_lottery_waiting",
+    "initial_prompt_review",
+    "growing",
+    "backlog_ranking",
+    "ranking",
+    "ready_for_export",
+    "aborted_low_grade",
+    "halted_by_moderator",
+  ];
+
+  const languages = [];
+  for (let i = 0; i < stat.stats.length; i++) {
+    if (!languages.includes(stat.stats[i].lang)) {
+      languages.push(stat.stats[i].lang);
+    }
+  }
+
+  const rows = [];
+
+  for (let i = 0; i < stat.stats.length; i++) {
+    const lang = stat.stats[i].lang;
+
+    let langRow = findLangRow(rows, lang);
+
+    if (langRow === -1) {
+      // The language is not yet in the data. Add it
+      rows.push(Array(headers.length).fill(0));
+      langRow = rows.length - 1;
+      rows[langRow][0] = lang;
+    }
+
+    const state = stat.stats[i].state;
+    const stateCol = headers.indexOf(state);
+    if (stateCol === -1) {
+      continue;
+    }
+
+    const threadCount = stat.stats[i].count;
+    rows[langRow][stateCol] = threadCount;
+  }
+
+  headers = headers.map((item) => t(getTypeSafei18nKey(item)));
+
+  for (let i = 0; i < rows.length; i++) {
+    const langCode = rows[i][0];
+    rows[i][0] = getLocaleDisplayName(langCode) + " [" + langCode + "]";
+  }
+  return { headers, rows };
+}
+
 export const MessageTreeStateStats = ({ stat, titleFn }: MessageTreeStateStatsProps) => {
   const { t } = useTranslation("stats");
+  const color = useColorModeValue(colors.light.text, colors.dark.text);
+  const langs = stat.stats.map((item) => item.lang);
+  const uniqueLangs = langs.filter((item, index) => langs.indexOf(item) === index);
+  const [selectedLang, setSelectedLang] = useState("en");
+  const stats = stat.stats.filter((item) => item.lang === selectedLang);
```

Comment:
can you `useMemo` on some of these values?

Reply:
I changed the code to use the `Set` approach to find the unique languages and states. In addition, I added useMemo. It's my first encounter with this hook, so I hope I did it ok!